ksql-engine/src/main/java/io/confluent/ksql/internal/KsqlMetricsExtension.java

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright 2019 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License (the "License"); you may not use
+ * this file except in compliance with the License.  You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.internal;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.function.Supplier;
+import org.apache.kafka.common.Configurable;
+import org.apache.kafka.common.metrics.MeasurableStat;
+
+/**
+ * This interface provides a way for users to provide custom metrics that will be emitted alongside
+ * the default KSQL engine JMX metrics.
+ */
+public interface KsqlMetricsExtension extends Configurable {
+
+  @Override
+  void configure(Map<String, ?> config);
+
+  /**
+   * Returns custom metrics to be emitted alongside the default KSQL engine JMX metrics.
+   *
+   * @return list of metrics
+   */
+  List<Metric> getCustomMetrics();
+
+  class Metric {
```

Comment:
should our internal metrics use this class? that way we would have something like `List<Metric> getSystemDefinedMetrics()` and just register them the same way we register the extensions (makes sure they both go through the exact same code paths)

Reply:
Done, except it's not as straight-forward as introducing `List<Metric> getSystemDefinedMetrics()` and looping through them to register the metrics, since many of the sensors need to be saved (see [here](https://github.com/confluentinc/ksql/blob/db2aab5802aa933dbab0f6dab53ef6bcccd6c153/ksql-engine/src/main/java/io/confluent/ksql/internal/KsqlEngineMetrics.java#L88-L94)). Curious to know whether you think the changes I've introduced are an improvement or not.